pkg/core/candidate/requestor.go

```diff
@@ -0,0 +1,120 @@
+package candidate
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/config"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/rpcbus"
+	lg "github.com/sirupsen/logrus"
+)
+
+var log = lg.WithField("process", "candidate-requestor")
+
+// Requestor serves to retrieve certain Candidate messages from peers in the
+// network.
+type Requestor struct {
+	lock                 sync.RWMutex
+	requestHash          []byte
+	publisher            eventbus.Publisher
+	requestCandidateChan <-chan rpcbus.Request
+	candidateQueue       chan message.Candidate
+}
+
+// NewRequestor returns an initialized Requestor struct.
+func NewRequestor(publisher eventbus.Publisher, rb *rpcbus.RPCBus) *Requestor {
+	requestCandidateChan := make(chan rpcbus.Request, 1)
+	if err := rb.Register(topics.GetCandidate, requestCandidateChan); err != nil {
+		panic(err)
+	}
+
+	return &Requestor{
+		requestHash:          make([]byte, 32),
+		publisher:            publisher,
+		requestCandidateChan: requestCandidateChan,
+		candidateQueue:       make(chan message.Candidate, 100),
+	}
+}
+
+// Listen allows the Requestor to receive incoming candidate requests and
+// process them accordingly. Should be ran in a goroutine.
+func (r *Requestor) Listen(ctx context.Context) {
+	for {
+		select {
+		case m := <-r.requestCandidateChan:
+			cm, err := r.requestCandidate(m)
+			if err != nil {
+				m.RespChan <- rpcbus.NewResponse(nil, err)
+				continue
+			}
+
+			m.RespChan <- rpcbus.NewResponse(cm, nil)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
```

Comment:
Remove this, pass the `Requestor` directly to the consensus and let the consensus steps call directly `Requestor.RequestCandidate`